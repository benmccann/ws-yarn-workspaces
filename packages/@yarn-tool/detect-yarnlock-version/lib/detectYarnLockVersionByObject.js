"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkV1 = exports.checkV2 = exports.detectYarnLockVersionByObject = void 0;
const types_1 = require("./types");
function detectYarnLockVersionByObject(yarnLockObject) {
    var _a, _b;
    if (((_b = (_a = yarnLockObject.__metadata) === null || _a === void 0 ? void 0 : _a.version) === null || _b === void 0 ? void 0 : _b.toString()) === '4' && checkV2(yarnLockObject)) {
        return types_1.EnumDetectYarnLock.berry;
    }
    else if (typeof yarnLockObject.type === 'string' && yarnLockObject.object && checkV1(yarnLockObject.object)) {
        return types_1.EnumDetectYarnLock.v1;
    }
    else if (!('type' in yarnLockObject) && !('object' in yarnLockObject) && checkV1(yarnLockObject)) {
        return types_1.EnumDetectYarnLock.v1;
    }
    return types_1.EnumDetectYarnLock.unknown;
}
exports.detectYarnLockVersionByObject = detectYarnLockVersionByObject;
function checkV2(obj) {
    let ks = Object.keys(obj)
        .slice(2)
        .filter(k => k !== '__metadata');
    let k = ks[0];
    if (ks.length && obj[k].version && obj[k].resolution && obj[k].linkType) {
        return types_1.EnumDetectYarnLock.berry;
    }
}
exports.checkV2 = checkV2;
function checkV1(obj) {
    let ks = Object.keys(obj)
        .slice(2);
    let k = ks[0];
    if (ks.length && obj[k].version && obj[k].resolved) {
        return types_1.EnumDetectYarnLock.v1;
    }
}
exports.checkV1 = checkV1;
exports.default = detectYarnLockVersionByObject;
//# sourceMappingURL=detectYarnLockVersionByObject.js.map