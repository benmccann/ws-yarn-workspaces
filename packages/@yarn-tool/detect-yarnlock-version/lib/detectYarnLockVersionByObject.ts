import { EnumDetectYarnLock } from './types';

export function detectYarnLockVersionByObject(yarnLockObject: Record<string, any>)
{
	if (yarnLockObject.__metadata?.version?.toString() === '4' && checkV2(yarnLockObject))
	{
		return EnumDetectYarnLock.berry
	}
	else if (typeof yarnLockObject.type === 'string' && yarnLockObject.object && checkV1(yarnLockObject.object))
	{
		return EnumDetectYarnLock.v1
	}
	else if (!('type' in yarnLockObject) && !('object' in yarnLockObject) && checkV1(yarnLockObject))
	{
		return EnumDetectYarnLock.v1
	}

	return EnumDetectYarnLock.unknown
}

export function checkV2(obj)
{
	let ks = Object.keys(obj)
		.slice(2)
		.filter(k => k !== '__metadata')
	;

	let k = ks[0];

	if (ks.length && obj[k].version && obj[k].resolution && obj[k].linkType)
	{
		return EnumDetectYarnLock.berry
	}
}

export function checkV1(obj)
{
	let ks = Object.keys(obj)
		.slice(2)
	;

	let k = ks[0];

	if (ks.length && obj[k].version && obj[k].resolved)
	{
		return EnumDetectYarnLock.v1
	}
}

export default detectYarnLockVersionByObject
