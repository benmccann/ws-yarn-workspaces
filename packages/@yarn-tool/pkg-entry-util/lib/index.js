"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pkgExportsVerify = exports.pkgExportsAddPJsonEntry = void 0;
var exports_1 = require("./field/exports");
Object.defineProperty(exports, "pkgExportsAddPJsonEntry", { enumerable: true, get: function () { return exports_1.pkgExportsAddPJsonEntry; } });
Object.defineProperty(exports, "pkgExportsVerify", { enumerable: true, get: function () { return exports_1.pkgExportsVerify; } });
//# sourceMappingURL=index.js.map