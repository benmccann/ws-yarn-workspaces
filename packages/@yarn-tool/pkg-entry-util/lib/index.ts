export { fixPkgBinField } from './field/bin';

export { fixPublishConfig } from './field/publishConfig';

export { pkgExportsAddPJsonEntry, pkgExportsVerify } from './field/exports';
