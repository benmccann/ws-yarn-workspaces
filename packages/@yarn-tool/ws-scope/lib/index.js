"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkspacesScope = void 0;
const find_root_1 = require("@yarn-tool/find-root");
const upath2_1 = require("upath2");
const json_1 = require("./format/json");
const yaml_1 = require("./format/yaml");
const array_hyper_unique_1 = require("array-hyper-unique");
const check_scope_1 = require("./util/check-scope");
const sort_lerna_json_1 = require("@yarn-tool/sort-lerna-json");
class WorkspacesScope {
    constructor(cwd) {
        this.rootData = (0, find_root_1.findRootLazy)({
            cwd,
            throwError: true,
            shouldHasWorkspaces: true,
        });
        this._root_package_json = new json_1.ScopeJson(this.resolvePath('package.json'), {
            field: 'workspaces',
        });
        this._root_lerna_json = new json_1.ScopeJson(this.resolvePath('lerna.json'), {
            field: 'packages',
        });
        this._root_pnpm_workspace_yaml = new yaml_1.ScopeYaml(this.resolvePath('pnpm-workspace.yaml'), {
            field: 'packages',
        });
        this._root_package_json.loadFileLazy();
        this._root_lerna_json.loadFileLazy();
        this._root_pnpm_workspace_yaml.loadFileLazy();
    }
    get changed() {
        return this._root_package_json.changed || this._root_lerna_json.changed || this._root_pnpm_workspace_yaml.changed;
    }
    resolvePath(...paths) {
        if (!paths.length) {
            throw new SyntaxError(`Invalid arguments: ${paths}`);
        }
        return (0, upath2_1.resolve)(this.rootData.ws, ...paths);
    }
    add(scope) {
        if (scope === (0, upath2_1.basename)(scope)) {
            scope = `packages/${scope}/*`;
        }
        (0, check_scope_1.assertScopePath)(scope, this.rootData.ws);
        this._root_package_json.addLazy(scope);
        this._root_lerna_json.addLazy(scope);
        this._root_pnpm_workspace_yaml.addLazy(scope);
        return scope;
    }
    remove(scope) {
        if (scope === (0, upath2_1.basename)(scope)) {
            scope = `packages/${scope}/*`;
        }
        (0, check_scope_1.assertScopePath)(scope, this.rootData.ws);
        this._root_package_json.removeLazy(scope);
        this._root_lerna_json.removeLazy(scope);
        this._root_pnpm_workspace_yaml.removeLazy(scope);
        return scope;
    }
    save() {
        if (this._root_lerna_json.opened) {
            (0, sort_lerna_json_1.sortLernaJson)(this._root_lerna_json.json);
        }
        this._root_package_json.saveFile();
        this._root_lerna_json.saveFile();
        this._root_pnpm_workspace_yaml.saveFile();
    }
    get value() {
        const value = [
            this._root_package_json.value,
            this._root_lerna_json.value,
            this._root_pnpm_workspace_yaml.value,
        ].flat().filter(v => v === null || v === void 0 ? void 0 : v.length);
        return (0, array_hyper_unique_1.array_unique_overwrite)(value);
    }
    syncValue() {
        const value = this.value
            .filter(v => (v === null || v === void 0 ? void 0 : v.length) && !v.startsWith('!'));
        (0, array_hyper_unique_1.array_unique_overwrite)(value);
        value.forEach(scope => this.add(scope));
        return value;
    }
}
exports.WorkspacesScope = WorkspacesScope;
exports.default = WorkspacesScope;
//# sourceMappingURL=index.js.map