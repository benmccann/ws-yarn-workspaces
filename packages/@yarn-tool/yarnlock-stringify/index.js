"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.yarnLockStringify = yarnLockStringify;
exports._yarnLockStringifyRaw = _yarnLockStringifyRaw;
exports.yarnLockStringifyRaw = yarnLockStringifyRaw;
exports._yarnLockStringifyRawCore = _yarnLockStringifyRawCore;
const detectYarnLockVersionByObject_1 = require("@yarn-tool/detect-yarnlock-version/lib/detectYarnLockVersionByObject");
const detectYarnLockVersion_1 = require("@yarn-tool/detect-yarnlock-version/lib/detectYarnLockVersion");
const yarnlock_error_1 = require("@yarn-tool/yarnlock-error");
const yarnlock_parsed_to_json_1 = require("@yarn-tool/yarnlock-parsed-to-json");
const yarnlock_types_1 = require("@yarn-tool/yarnlock-types");
const detect_yarnlock_version_1 = require("@yarn-tool/detect-yarnlock-version");
const v1_1 = require("@yarn-tool/yarnlock-parse-raw/lib/v1");
const v2_1 = require("@yarn-tool/yarnlock-parse-raw/lib/v2");
function yarnLockStringify(yarnlock_old) {
    let verType = (0, detect_yarnlock_version_1.detectYarnLockVersionByParsed)(yarnlock_old);
    if (verType) {
        yarnlock_old = (0, yarnlock_parsed_to_json_1.yarnLockParsedToRawJSON)(yarnlock_old);
    }
    else {
        verType = (0, detectYarnLockVersionByObject_1.detectYarnLockVersionByObject)(yarnlock_old);
    }
    if (verType) {
        return _yarnLockStringifyRawCore(verType, yarnlock_old).content;
    }
    else {
        verType = (0, detectYarnLockVersion_1.detectYarnLockVersion)(yarnlock_old);
        if (verType) {
            return yarnlock_old.toString();
        }
    }
    throw (0, yarnlock_error_1.newYarnLockParsedVersionError)();
}
function _yarnLockStringifyRaw(yarnlockRawJSON) {
    const verType = (0, detectYarnLockVersionByObject_1.detectYarnLockVersionByObject)(yarnlockRawJSON);
    return _yarnLockStringifyRawCore(verType, yarnlockRawJSON);
}
function yarnLockStringifyRaw(yarnlockRawJSON) {
    return _yarnLockStringifyRaw(yarnlockRawJSON).content;
}
function _yarnLockStringifyRawCore(verType, yarnlockRawJSON) {
    let fn;
    switch (verType) {
        case yarnlock_types_1.EnumDetectYarnLock.v3:
        case yarnlock_types_1.EnumDetectYarnLock.v2:
            fn = v2_1.stringifyYarnLockRawV2;
            break;
        case yarnlock_types_1.EnumDetectYarnLock.v1:
            fn = v1_1.stringifyYarnLockRawV1;
            break;
        default:
            throw (0, yarnlock_error_1.newYarnLockParsedVersionError)();
    }
    return {
        verType,
        fn,
        json: yarnlockRawJSON,
        content: fn(yarnlockRawJSON),
    };
}
exports.default = yarnLockStringify;
//# sourceMappingURL=index.js.map