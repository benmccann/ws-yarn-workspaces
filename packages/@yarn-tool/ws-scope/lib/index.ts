import { findRootLazy, IFindRootReturnType } from '@yarn-tool/find-root';
import { resolve } from 'upath2';
import { ScopeJson } from './format/json';
import { ScopeYaml } from './format/yaml';
import { array_unique_overwrite } from 'array-hyper-unique';

export class WorkspacesScope
{
	rootData: IFindRootReturnType;

	_root_package_json: ScopeJson<'workspaces'>;
	_root_lerna_json: ScopeJson<'packages'>;
	_root_pnpm_workspace_yaml: ScopeJson<'packages'>;

	constructor(cwd?: string)
	{
		this.rootData = findRootLazy({
			cwd,
			throwError: true,
			shouldHasWorkspaces: true,
		});

		this._root_package_json = new ScopeJson(this.resolvePath('package.json'), {
			field: 'workspaces',
		});
		this._root_lerna_json = new ScopeJson(this.resolvePath('lerna.json'), {
			field: 'packages',
		});
		this._root_pnpm_workspace_yaml = new ScopeYaml(this.resolvePath('pnpm-workspace.yaml'), {
			field: 'packages',
		});

		this._root_package_json.existsFile() && this._root_package_json.loadFile();
		this._root_lerna_json.existsFile() && this._root_lerna_json.loadFile();
		this._root_pnpm_workspace_yaml.existsFile() && this._root_pnpm_workspace_yaml.loadFile();
	}

	get changed()
	{
		return this._root_package_json.changed || this._root_lerna_json.changed || this._root_pnpm_workspace_yaml.changed;
	}

	resolvePath(...paths: [string, ...string[]])
	{
		if (!paths.length)
		{
			throw new SyntaxError(`Invalid arguments: ${paths}`)
		}

		return resolve(this.rootData.ws, ...paths);
	}

	add(scope: string)
	{
		this._root_package_json.add(scope);
		this._root_lerna_json.add(scope);
		this._root_pnpm_workspace_yaml.add(scope);
	}

	remove(scope: string)
	{
		this._root_package_json.remove(scope);
		this._root_lerna_json.remove(scope);
		this._root_pnpm_workspace_yaml.remove(scope);
	}

	save()
	{
		this._root_package_json.saveFile();
		this._root_lerna_json.saveFile();
		this._root_pnpm_workspace_yaml.saveFile();
	}

	get value()
	{
		const value = [
			this._root_package_json.value,
			this._root_lerna_json.value,
			this._root_pnpm_workspace_yaml.value,
		].flat().filter(v => v?.length)

		return array_unique_overwrite(value)
	}

	syncValue()
	{
		const value = this.value
			.filter(v => v?.length && !v.startsWith('!'))
		;

		array_unique_overwrite(value);

		value.forEach(scope => this.add(scope));

		return value
	}

}

export default WorkspacesScope
