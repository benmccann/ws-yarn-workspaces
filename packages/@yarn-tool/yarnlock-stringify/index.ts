import { detectYarnLockVersionByObject } from '@yarn-tool/detect-yarnlock-version/lib/detectYarnLockVersionByObject';
import { detectYarnLockVersion } from '@yarn-tool/detect-yarnlock-version/lib/detectYarnLockVersion';
import { newYarnLockParsedVersionError } from '@yarn-tool/yarnlock-error';
import { yarnLockParsedToRawJSON } from '@yarn-tool/yarnlock-parsed-to-json';
import {
	EnumDetectYarnLock,
	IYarnLockParsed,
	IYarnLockRawSourceV1,
	IYarnLockRawSourceV2,
	IYarnLockSource,
} from '@yarn-tool/yarnlock-types';
import { detectYarnLockVersionByParsed } from '@yarn-tool/detect-yarnlock-version';
import { stringifyYarnLockRawV1 } from '@yarn-tool/yarnlock-parse-raw/lib/v1';
import { stringifyYarnLockRawV2 } from '@yarn-tool/yarnlock-parse-raw/lib/v2';

export function yarnLockStringify(yarnlock_old: Record<string, any> | Buffer | string | IYarnLockSource | IYarnLockParsed): string
{
	let verType: EnumDetectYarnLock = detectYarnLockVersionByParsed(yarnlock_old as IYarnLockParsed);

	if (verType)
	{
		yarnlock_old = yarnLockParsedToRawJSON(yarnlock_old as IYarnLockParsed)
	}
	else
	{
		verType = detectYarnLockVersionByObject(yarnlock_old as any);
	}

	if (verType)
	{
		return _yarnLockStringifyRawCore(verType, yarnlock_old as IYarnLockSource).content
	}
	else
	{
		verType = detectYarnLockVersion(yarnlock_old as string);

		if (verType)
		{
			return yarnlock_old.toString()
		}
	}

	throw newYarnLockParsedVersionError()
}

export type IResultYarnLockStringifyRaw = {
	verType: EnumDetectYarnLock.v1;
	fn: typeof stringifyYarnLockRawV1
	json: IYarnLockRawSourceV1;
	content: string;
} | {
	verType: EnumDetectYarnLock.v2;
	fn: typeof stringifyYarnLockRawV2
	json: IYarnLockRawSourceV2;
	content: string;
} | {
	verType: EnumDetectYarnLock.v3;
	fn: typeof stringifyYarnLockRawV2
	json: IYarnLockRawSourceV2;
	content: string;
};

export function _yarnLockStringifyRaw(yarnlockRawJSON: IYarnLockSource)
{
	const verType = detectYarnLockVersionByObject(yarnlockRawJSON)

	return _yarnLockStringifyRawCore(verType, yarnlockRawJSON)
}

export function yarnLockStringifyRaw(yarnlockRawJSON: IYarnLockSource)
{
	return _yarnLockStringifyRaw(yarnlockRawJSON).content
}

export function _yarnLockStringifyRawCore(verType: EnumDetectYarnLock, yarnlockRawJSON: IYarnLockSource): IResultYarnLockStringifyRaw
{
	let fn: typeof stringifyYarnLockRawV2 | typeof stringifyYarnLockRawV1;

	switch (verType)
	{
		case EnumDetectYarnLock.v3:
		case EnumDetectYarnLock.v2:
			fn = stringifyYarnLockRawV2;
			break;
		case EnumDetectYarnLock.v1:
			fn = stringifyYarnLockRawV1;
			break;
		default:
			throw newYarnLockParsedVersionError()
	}

	return {
		verType,
		fn,
		json: yarnlockRawJSON,
		content: fn(yarnlockRawJSON),
	} as IResultYarnLockStringifyRaw
}

export default yarnLockStringify
