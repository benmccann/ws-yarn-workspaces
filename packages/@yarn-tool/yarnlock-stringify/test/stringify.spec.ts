//@noUnusedParameters:false

import { basename, dirname, extname, join } from 'path';
import {
	detectYarnLockVersion,
	_tryParse,
	detectYarnLockVersionByDir,
	detectYarnLockVersionByFile,
	detectYarnLockVersionByParsed,
} from '@yarn-tool/detect-yarnlock-version';
import { __TEST_YARNLOCK } from '../../../../__root_ws';
import { sync as FastGlob } from '@bluelovers/fast-glob';
import { readFileSync } from 'fs';
import { yarnLockParse } from '@yarn-tool/yarnlock-parse';
import { yarnLockStringify } from '../index';
import { detectYarnLockVersionByObject } from '@yarn-tool/detect-yarnlock-version/lib/detectYarnLockVersionByObject';
import { EnumDetectYarnLock } from '@yarn-tool/yarnlock-types';

beforeAll(async () =>
{

});

describe(`version`, () =>
{

	(<(keyof typeof EnumDetectYarnLock)[]>[
		'v1',
		'v2',
		'v3',
	]).forEach(ver =>
	{
		describe(ver, () =>
		{
			const dir = join(__TEST_YARNLOCK, ver);

			FastGlob([
				'**/*.lock',
			], {
				cwd: dir,
			}).forEach(name =>
			{
				const file = join(dir, name);
				const buf = readFileSync(file);

				const rawObject = yarnLockParse(buf);

				test(name, () =>
				{

					let actual = yarnLockStringify(rawObject);
					let expected = EnumDetectYarnLock[ver];

					expect(detectYarnLockVersion(actual)).toStrictEqual(expected);

					const rawObject2 = yarnLockParse(actual);

					expect(detectYarnLockVersionByParsed(rawObject2)).toStrictEqual(expected);

					let actual2 = yarnLockStringify(rawObject2);

					expect(rawObject2).toStrictEqual(rawObject);
					expect(actual2).toStrictEqual(actual);

					expect(actual.slice(0, 160)).toMatchSnapshot();

				});
			})
		})

	});

})
