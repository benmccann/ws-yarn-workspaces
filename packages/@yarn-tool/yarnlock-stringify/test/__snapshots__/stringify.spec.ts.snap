// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`version v1 yarn.lock 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


\\"@babel/code-frame@^7.0.0\\", \\"@babel/code-frame@^7.10.1\\":
  version \\"7.10.1\\"
  resolved \\"https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.10.1.tgz#d5481c5095daa1c57e16e54c6f9198443afb49ff\\"
  integrity sha512-IGhtTmpjGbYzcEDOw7DcQtbQSXcG9ftmAXtWTu9V936vDye4xjjekktF"
`;

exports[`version v2 meta-only.lock 1`] = `
"# This file is generated by running \\"yarn install\\" inside your project.
# Manual changes might be lost - proceed with caution!

__metadata:
  version: 4
"
`;

exports[`version v2 yarn.lock 1`] = `
"# This file is generated by running \\"yarn install\\" inside your project.
# Manual changes might be lost - proceed with caution!

__metadata:
  version: 4

\\"library@npm:>=1.0.0, library@npm:>=1.1.0\\":
  version: 3.0.0
  resolution: \\"library@npm:3.0.0\\"
  checksum: 2/6fc98369f954229cbb6e4ce1a27a4202a17b4ad67b8def64322e9351414eef405b8783ecbe1b8f0334e52935ae8c43cd3"
`;

exports[`version v3 meta-only.lock 1`] = `
"# This file is generated by running \\"yarn install\\" inside your project.
# Manual changes might be lost - proceed with caution!

__metadata:
  version: 6
  cacheKey: 8
"
`;

exports[`version v3 yarn.lock 1`] = `
"# This file is generated by running \\"yarn install\\" inside your project.
# Manual changes might be lost - proceed with caution!

__metadata:
  version: 6
  cacheKey: 8

\\"@bluelovers/fast-glob@npm:^3.0.4\\":
  version: 3.0.4
  resolution: \\"@bluelovers/fast-glob@npm:3.0.4\\"
  dependencies:
    bluebird: ^3
    fast-glob: ^3
  checksum: b7417447ae8fa5c1cde2dded1f11"
`;

exports[`version v3 yarn-version-7.lock 1`] = `
"# This file is generated by running \\"yarn install\\" inside your project.
# Manual changes might be lost - proceed with caution!

__metadata:
  version: 7
  cacheKey: 9

\\"@actions/core@npm:^1.2.6\\":
  version: 1.2.6
  resolution: \\"@actions/core@npm:1.2.6\\"
  checksum: 034e57fcb5f944d5fb0ef55be1b212dd88e23d1a50aaffda874cb94e8f4bfa633a66f108f26e81a7cce287cd2b349aa"
`;
